Source/ISeriesProvider/DB2iSeriesDataProvider.cs

```diff
@@ -141,70 +144,183 @@ private void SetupAccessClient()
-			SetProviderField(adapter.iDB2DataLinkType, typeof(string), adapter.GetiDB2DataLinkReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DateType, typeof(DateTime), adapter.GetiDB2DateReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DbClobType, typeof(string), adapter.GetiDB2DbClobReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DecFloat16Type, typeof(decimal), adapter.GetiDB2DecFloat16ReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DecFloat34Type, typeof(decimal), adapter.GetiDB2DecFloat34ReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DecimalType, typeof(decimal), adapter.GetiDB2DecimalReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2DoubleType, typeof(double), adapter.GetiDB2DoubleReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2GraphicType, typeof(string), adapter.GetiDB2GraphicReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2IntegerType, typeof(int), adapter.GetiDB2IntegerReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2NumericType, typeof(decimal), adapter.GetiDB2NumericReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2RealType, typeof(float), adapter.GetiDB2RealReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2RowidType, typeof(byte[]), adapter.GetiDB2RowidReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2SmallIntType, typeof(short), adapter.GetiDB2SmallIntReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2TimeType, typeof(DateTime), adapter.GetiDB2TimeReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2TimeStampType, typeof(DateTime), adapter.GetiDB2TimeStampReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2VarBinaryType, typeof(byte[]), adapter.GetiDB2VarBinaryReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2VarCharType, typeof(string), adapter.GetiDB2VarCharReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2VarCharBitDataType, typeof(byte[]), adapter.GetiDB2VarCharBitDataReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2VarGraphicType, typeof(string), adapter.GetiDB2VarGraphicReaderMethod, dataReaderType: adapter.DataReaderType);
-			SetProviderField(adapter.iDB2XmlType, typeof(string), adapter.GetiDB2XmlReaderMethod, dataReaderType: adapter.DataReaderType);
+			SetProviderField(typeof(long), adapter.iDB2BigIntType, adapter.GetiDB2BigIntReaderMethod, dataReaderType: adapter.DataReaderType);
```

Comment:
I struggled with this a little myself. I referenced linq2db and debugged the code by hand. The SetProviderField method creates a SQL deserializer of sorts. linq2db looks up which deserializer to use based on several fields. Ultimately, the deserializers setup here were never used because the fields were switched. When I swapped them, they suddenly started getting used. 

Here is where the provider field is getting set. Notice that the field type actually sets the provider field type:
https://github.com/linq2db/linq2db/blob/5dec345f6d597310c808d369f68fb88322c80f92/Source/LinqToDB/DataProvider/DynamicDataProviderBase.cs#L145

Here is where they are read:
https://github.com/linq2db/linq2db/blob/5dec345f6d597310c808d369f68fb88322c80f92/Source/LinqToDB/DataProvider/DataProviderBase.cs#L226

Reply:
Well there's serialization and deserialization, so perhaps the intent was different. Let's try switching back and see what problems it causes.

Reply:
Sure. Let's take a look at this when we meet next.